### ISSUE TITLE
Errors with table having a virtual column

### ISSUE THREAD

**Issue by rodydavis** [simolus3/moor#2097]
Example code:

https://github.com/rodydavis/flutter_sqlite_graph_db

<img width="969" alt="Screen Shot 2022-10-10 at 2 00 02 AM" src="https://user-images.githubusercontent.com/31253215/194831523-e5e286ea-9a31-488c-8386-76b04f151846.png">

The issue seems to be in the building of the SQL files, there are a lot of errors, but the columns are valid.

**Comment by rodydavis** [comment]
For reference I am implementing a graph database as described here:

https://github.com/dpapathanasiou/simple-graph

**Comment by rodydavis** [comment]
For example:

```
[INFO] drift_dev on lib/database/database.dart:line 3, column 37: Some columns are required but not present here. Expected values for id.
  ╷
3 │ insertNode: INSERT INTO nodes VALUES(json(?));
```

The schema defines a virtual id:

```sql
CREATE TABLE IF NOT EXISTS nodes (
    body TEXT,
    id   TEXT GENERATED ALWAYS AS (json_extract(body, '$.id')) VIRTUAL NOT NULL UNIQUE
);
```

**Comment by simolus3** [comment]
> Some columns are required but not present here.

This is a bug in `drift_dev`, it didn't consider that the column is generated. I have fixed this in 3d969c787f90377bf2cbfe148626f92b8918792b.

With queries like the following, I think the error is somewhat understandable:

```sql
traverseInbound: WITH RECURSIVE traverse(id) AS (
  SELECT :source
  UNION
  SELECT source FROM edges JOIN traverse ON target = id
) SELECT id FROM traverse;
```

You are selecting from a variable, but drift has no idea what type this variable has at this moment, which is what it's complaining about here. For efficiency, type inference in recursive CTEs is forward-only, drift is not smart enough to infer that, since `edges.source` is text, `:source` must be text too.

Luckily, you can fix that by giving the variable an explicit type annotation:

```sql
traverseInbound(:source AS TEXT): WITH RECURSIVE traverse(id) AS
```

You'll have to do that for every query constructing a recursive lookup like that.

**Comment by rodydavis** [comment]
Wow this is so helpful! I’ll update my example too. Working on a blog post about it 👍🏼

thank you so much for the fast turn around!
